deeprank/features/ResidueDensity.py

```diff
@@ -42,6 +47,12 @@ def get(self,cutoff=5.5):
                                            cutoff = cutoff,
                                            return_contact_pairs=True)
 
+        if len(res) < 5:
```

Comment:
why lower than 5 ?

Reply:
Protein-protein interfaces are quite large. If the number of contact residue pairs is smaller than 5, this complex is almost for sure a wrong model. We do not need such complexes in our training or testing data.

Reply:
maybe sometimes we want to have wrong conformation in the training set. I find that a bit arbitrary. 
It's always a good practice to have a given class doing one single thing. I'm not respecting this guideline all the time myself, but it's good practice

The purpose of this class is to compute the residue density of a given conformation. But now the class can also decides which conformation to exclude from the dataset ... so it does 2 things

If the issue arises when there is no contact residue I would prefer to just fix that instead of introducing complexity in the code by allowing the class to exclude conformation

Reply:
We indeed want wrong conformations in our training set but not those ones :), because a model with such a small interface can be easily identified as a wrong model by simply looking at it.

Thanks for the advice of designing of the software. Shall I write a function for it so that the user can choose to remove such models or not?

This "lower than 5 " condition will not fix the no contact error message, in fact. ResidueDensity.py is trying to count different types of contacts, such as "polar-polar", "apolar-polar" and so on. When an interface does not have any "polar-polar" interactions, the `extract_features()` function will return an empy dictionary for this key, and will cause an error message in  `export_data_hdf5()` in this line `ds = np.array(ds).astype('|S'+str(len(ds[0])))` in `FeatureClass.py`. To avoid an error break here, I modified `FeatureClass.py` by adding

```
            if len(ds) ==0 :
                self.error = True
              return 

```

Reply:
Here are some models for reproduce this bug: 
2I25/2I25_ranair-itw_237w.pdb
2I25/2I25_ranair-it1_155.pdb
2I25/2I25_ranair-it1_387.pdb
2I25/2I25_ranair-itw_387w.pdb
2I25/2I25_cm-it0_136.pdb
2I25/2I25_ranair-it1_155.pdb
2I25/2I25_ranair-it0_2784.pdb
2I25/2I25_ranair-itw_237w.pdb
2I25/2I25_cm-it0_3183.pdb
2I25/2I25_ranair-itw_387w.pdb
2I25/2I25_cm-it0_2624.pdb
2I25/2I25_ranair-it1_387.pdb

Reply:
As proposed on Slack we could have a preprocessing check for each PDB on the quality of the file (CB missing etc) and the quality of the interface (discard interface with less than 5 contact residues
)

Reply:
Ok, "discard interface with less than 5 contact residues" can be solved with the preprocessing check.

But we still have a separate problem (sorry for mixing two problems into one issue):

`ResidueDensity.py` is trying to count different types of contacts, such as "polar-polar", "apolar-polar" and so on. When an interface does not have any "polar-polar" interactions, for example, the `extract_features()` function will return an empy dictionary for this key, and will cause an error message in `export_data_hdf5()` in this line `ds = np.array(ds).astype('|S'+str(len(ds[0])))` in `FeatureClass.py`. To avoid an error break here, I modified FeatureClass.py by adding

```
            if len(ds) ==0 :
                self.error = True
              return 
```

Reply:
when `len(ds)==0`, maybe it is OK to set
`ds = np.array(ds)` (ds is empty list) instead of  `ds = np.array(ds).astype('|S'+str(len(ds[0])))`.
I'm not sure the empty feature will give problem or not for next steps of mapping and learning.

Reply:
Once we have `len(ds)==0` error, deeprank will delete this model from hdf5 file. Therefore, it will not be used in mapping and training.